File: frontend/src/types/order.ts
export enum OrderStatus {
  PENDING = "pending",
  PROCESSING = "processing",
  SHIPPED = "shipped",
  DELIVERED = "delivered",
  CANCELLED = "cancelled"
}

export interface OrderItem {
  order_id: string;
  item_id: string;
  quantity: number;
  price_at_time: number;
}

export interface OrderItemCreate {
  item_id: string;
  quantity: number;
  price_at_time?: number;
}

export interface Order {
  id: string;
  user_id: string;
  status: OrderStatus;
  shipping_address: string | null;
  total_amount: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
  items: OrderItem[];
}

export interface OrderCreate {
  user_id: string;
  status?: OrderStatus;
  shipping_address?: string | null;
  total_amount: number;
  notes?: string | null;
  items: OrderItemCreate[];
}

export interface OrderUpdate {
  status?: OrderStatus;
  shipping_address?: string | null;
  notes?: string | null;
}

File: backend/app/schemas/order.py
from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.order import OrderStatus


class OrderItemBase(BaseModel):
    """Base schema for OrderItem data (中間テーブル)."""
    item_id: UUID
    quantity: int = Field(..., gt=0)
    price_at_time: float = Field(..., gt=0)


class OrderItemCreate(OrderItemBase):
    """Schema for creating a new order item."""
    pass


class OrderBase(BaseModel):
    """Base schema for Order data."""
    user_id: UUID
    status: OrderStatus = OrderStatus.PENDING
    shipping_address: Optional[str] = None
    total_amount: float = Field(..., gt=0)
    notes: Optional[str] = None


class OrderCreate(OrderBase):
    """Schema for creating a new order."""
    items: List[OrderItemCreate]


class OrderUpdate(BaseModel):
    """Schema for updating an order."""
    status: Optional[OrderStatus] = None
    shipping_address: Optional[str] = None
    notes: Optional[str] = None


# Import Item schema to use in response
from app.schemas.item import Item as ItemSchema


class OrderItem(OrderItemBase):
    """Schema for OrderItem data."""
    order_id: UUID
    item: Optional[ItemSchema] = None

    class Config:
        from_attributes = True


class OrderInDBBase(OrderBase):
    """Base schema for Order data in the database."""
    id: UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class Order(OrderInDBBase):
    """Schema for Order data with items."""
    items: List[ItemSchema] = []

    class Config:
        from_attributes = True